ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java

```diff
@@ -93,11 +95,60 @@
-    this(Type.JournalBased, offset, null, sessionId, incarnationId, inclusive);
+    this(Type.JournalBased, offset, null, sessionId, incarnationId, inclusive, VERSION_2);
   }
 
-  private StoreFindToken(Type type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId,
-      boolean inclusive) {
+  /**
+   * Uninitialized token with the given version. Refers to the starting of the log.
+   * @param version refers to the version of the token
+   */
+  private StoreFindToken(short version) {
+    this(Type.Uninitialized, null, null, null, null, true, version);
+  }
+
+  /**
+   * Index based token with the given version. Refers to an index segment start offset and a store key that belongs to
+   * that index segment
+   * @param key The {@link StoreKey} which the token refers to. Index segments are keyed on store keys and hence
+   * @param indexSegmentStartOffset the start offset of the index segment which the token refers to
+   * @param sessionId the sessionId of the store
+   * @param incarnationId the incarnationId of the store
+   * @param version refers to the version of the token
+   */
+  private StoreFindToken(StoreKey key, Offset indexSegmentStartOffset, UUID sessionId, UUID incarnationId,
+      short version) {
+    this(Type.IndexBased, indexSegmentStartOffset, key, sessionId, incarnationId, false, version);
+  }
+
+  /**
+   * Journal based token with the given version. Refers to an offset in the journal
+   * @param offset the offset that this token refers to in the journal
+   * @param sessionId the sessionId of the store
+   * @param incarnationId the incarnationId of the store
+   * @param inclusive {@code true} if the offset is inclusive or in other words the blob at the given offset is inclusive.
+   *                  {@code false} otherwise
+   * @param version refers to the version of the token
+   */
+  private StoreFindToken(Offset offset, UUID sessionId, UUID incarnationId, boolean inclusive, short version) {
+    this(Type.JournalBased, offset, null, sessionId, incarnationId, inclusive, version);
+  }
+
+  /**
+   * Instantiating {@link StoreFindToken}
+   * @param type the {@link Type} of the token
+   * @param offset the offset that this token refers to
+   * @param key The {@link StoreKey} which the token refers to. Index segments are keyed on store keys and hence
+   * @param sessionId the sessionId of the store that this token referes to
+   * @param incarnationId the incarnationId of the store that this token refers to
+   * @param inclusive {@code true} if the offset is inclusive or in other words the blob at the given offset is inclusive.
+   *                  {@code false} otherwise
+   * @param version refers to the version of the token
+   */
+  private StoreFindToken(Type type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId, boolean inclusive,
+      short version) {
+    if (version != VERSION_0 && version != VERSION_1 && version != VERSION_2) {
```

Comment:
this is a private constructor - when can this situation happen? When constructing from constructors, the version is the current version by default. When deserializing, a wrong version would have failed deserialization.